### DISCUSSION TITLE
How to create a dynamics navigation with API

### DISCUSSION THREAD

**Discussion by DBonbon** [vercel/next.js#50911] [App Router]
I try to create a dynamic navigation component, I cannot find a proper way to use the API list which provides all pages information. 

Precisely, I'm looking to consume the date from the following:
```
export async function getAllPages() {
    return await getRequest(`${API_URL}/v1/page_relative_urls/`);
}
```
In my browser the `${API_URL}/v1/page_relative_urls/` displays the data as expected:

{ "meta": 
   { "total_count": 8 }, 
"items": [ 
   { 
   "title": "Home Page", 
   "relative_url": "/"
   }, 
  { "title": "About us",
  "relative_url": "/about-us/"
  }, 
 ... 
]}

I cannot find though how to call getAllPages. 
In components/Navigation I tried the following 

```
import getAllPages from '../../api/wagtail.js'

const Navtest = () => {
    return <div className={s.Root}>

   <nav>
        {getAllPages.items.map((link, index) => {
          return (
            <ul>
              <Link href={link.title}>
                <li key={index}>{link.title}</li>
              </Link>
            </ul>
          );
        })}
      </nav>

    </div>;
};

Navtest.propTypes = {};

Navtest.defaultProps = {};

export default Navtest;
```

This raises a red mark error below the getAllPages... which means that I don't call it properly. I sat on it for a couple of days and I'm still in the first square. I'm not even sure what is the right question, problem I need to address. 

Any return, tip would be highly appreciated.

**Comment by SamKomesarook** [comment]
`getAllPages` is a function, so you need to call it:
```typescript
{getAllPages().items.map((link, index) => {
          return (
            <ul>
              <Link href={link.title}>
                <li key={index}>{link.title}</li>
              </Link>
            </ul>
          );
        })}
```
At the moment, you're referencing the function object itself.

**Reply by DBonbon** [reply]
Thanks @SamKomesarook for your help. It didn'te work out for me. 

The following raises the error `Unhandled Runtime Error

TypeError: _api_wagtail_js__WEBPACK_IMPORTED_MODULE_1__["default"] is not a function`:

```const Navtest = () => {
    return <div className={s.Root}>

   <nav>
        {getAllPages().items.map((link, index) => {
          return (
            <ul>
              <Link href={link.title}>
                <li key={index}>{link.title}</li>
              </Link>
            </ul>
          );
        })}
      </nav>

    </div>;
};


If I add an curly braces to getAllPages, i.e. import {getAllPages} from '../../api/wagtail.js';


It raises the error TypeError: _api_wagtail_js__WEBPACK_IMPORTED_MODULE_1__.getAllPages().items is undefined

**Reply by SamKomesarook** [reply]
So I am not too sure about these webpack errs, but I missed that your `getAllPages` is async. Assuming this component is a client one, perhaps use the [use hook](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#use-in-client-components):
```typescript
"use client"
import getAllPages from '../../api/wagtail.js'
import { use } from "react"

const Navtest = () => {
const pages = use(getAllPages())
    return (<div className={s.Root}>
   <nav>
        {pages.items.map((link, index) => {
          return (
            <ul>
              <Link href={link.title}>
                <li key={index}>{link.title}</li>
              </Link>
            </ul>
          );
        })}
      </nav>
    </div>);
};

Navtest.propTypes = {};

Navtest.defaultProps = {};

export default Navtest;
```

**Reply by DBonbon** [reply]
Thanks @SamKomesarook for getting back. I tried, it still doesn't work. I'm not so sure to properly understood the hook's functionality here. The component is a server one. It is simply a menu that iterates through the items provided in the getAllPages.

**Reply by SamKomesarook** [reply]
If its a server component, try this:
```typescript
import getAllPages from '../../api/wagtail.js'

const Navtest = async () => {
const data = await getAllPages()

    return <div className={s.Root}>

   <nav>
        {data && data.items.map((link, index) => {
          return (
            <ul>
              <Link href={link.title}>
                <li key={index}>{link.title}</li>
              </Link>
            </ul>
          );
        })}
      </nav>

    </div>;
};

Navtest.propTypes = {};

Navtest.defaultProps = {};

export default Navtest;
``` 

I think a big issue is that your fetch is async, and the component that calls it needs to handle that by becoming an async component of its own, fetching the data outside of the render process, and then checking that the data is available before rendering.

**Comment by DBonbon** [comment]
Many thanks for the support. I add it as is. It raises the error:

"Server Error

Error: Objects are not valid as a React child (found: [object Promise]). If you meant to render a collection of children, use an array instead."

**Reply by SamKomesarook** [reply]
What is the return type of `getRequest`? Once its resolved, it should be an object, however it looks like its still a promise...

**Reply by DBonbon** [reply]
I'm not so sure how to answer this. 
I hope the following console logs can help. Kindly advise if I can provide any other information. 

The following command, run in the ArticlePage (which have the child component Navtest) shows:
console.log(xtype({getAllPages})) //single_prop_object

If I place the Navtest instead, it provides the same output, single_prop_object. Yet, only after the page is refreshed. 

In case it helps, before refreshing the the Navtest console.log shows 5 these entries:
 
console.log(xtype({Navtest}))

1.
 [Fast Refresh] rebuilding   //[hot-dev-client.js:185:24](webpack://_N_E/node_modules/next/dist/client/dev/error-overlay/hot-dev-client.js?1600)

2.
[Fast Refresh] done in 474ms  //[hot-dev-client.js:161:16](webpack://_N_E/node_modules/next/dist/client/dev/error-overlay/hot-dev-client.js?1600)

3.
single_prop_object //[ArticlePage.js:13:12](webpack://_N_E/containers/ArticlePage/ArticlePage.js?d6f5)

4. multi_prop_object // [ArticlePage.js:14:12](webpack://_N_E/containers/ArticlePage/ArticlePage.js?d6f5)

5. single_prop_object // [ArticlePage.js:15:12](webpack://_N_E/containers/ArticlePage/ArticlePage.js?d6f5)

**Comment by DBonbon** [comment]
I decide to close this discussion and open a new one, as I suspect, I didn't provide a proper information so the forum can properly locate the bugs in my code. I wish to provide credit to the members who did their best to help me. I don't wish however, to mark this thread as answer, since it is misleading. Kindly advise, if I can do so and/if I should re-open the discussion. The current question is in the following [https://github.com/vercel/next.js/discussions/51505](url)
